Claims Broken

When starting from Ethereum mainnet, and having to switch, the RPC stays broken until localStorage is cleared.

- On the production instance https://github.com/ubiquity/pay.ubq.fi/commit/da14f1b claims don't work with the following error:

```
erc20-permit.ts:259 Error in nonceBitmap method:  Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)
    at _Logger.makeError (index.ts:269:28)
    at _Logger.throwError (index.ts:281:20)
    at JsonRpcProvider.<anonymous> (json-rpc-provider.ts:483:23)
    at Generator.throw (<anonymous>)
    at rejected (index.ts:373:5)
(anonymous) @ erc20-permit.ts:259
Promise.catch (async)
isNonceClaimed @ erc20-permit.ts:258
checkPermitClaimable @ erc20-permit.ts:134
checkPermitClaimability @ erc20-permit.ts:49
claimErc20PermitHandler @ erc20-permit.ts:109
erc20-permit.ts:136 Error in isNonceClaimed:  Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)
    at _Logger.makeError (index.ts:269:28)
    at _Logger.throwError (index.ts:281:20)
    at JsonRpcProvider.<anonymous> (json-rpc-provider.ts:483:23)
    at Generator.throw (<anonymous>)
    at rejected (index.ts:373:5)
```

--- a/static/scripts/rewards/render-transaction/read-claim-data-from-url.ts
+++ b/static/scripts/rewards/render-transaction/read-claim-data-from-url.ts
@@ -31,7 +31,11 @@ export async function readClaimDataFromUrl(app: AppState) {
 
   app.claims = decodeClaimData(base64encodedTxData).flat();
   app.claimTxs = await getClaimedTxs(app);
-  app.provider = await useFastestRpc(app);
+  try {
+    app.provider = await useFastestRpc(app);
+  } catch (e) {
+    toaster.create("error", `${e}`);
+  }
   if (ethereum) {
     try {
       app.signer = await connectWallet();

--- a/static/scripts/rewards/rpc-optimization/get-fastest-rpc-provider.ts
+++ b/static/scripts/rewards/rpc-optimization/get-fastest-rpc-provider.ts
@@ -24,7 +24,7 @@ export async function getFastestRpcProvider(networkId: number) {
       await rpcProvider.getBlock(1);
       return rpcProvider;
     } catch (e) {
-      console.log(`Failed to get a block using network ${optimalRPC}, will try with another.`);
+      console.warn(`Failed to get a block using network ${optimalRPC}, will try with another.`);
       delete latencies[optimalRPC];
     }
   }

--- a/static/scripts/rewards/rpc-optimization/get-fastest-rpc-provider.ts
+++ b/static/scripts/rewards/rpc-optimization/get-fastest-rpc-provider.ts
@@ -1,18 +1,33 @@
 import { ethers } from "ethers";
 
-export function getFastestRpcProvider(networkId: number) {
+const MAX_RETRIES = 10;
+
+export async function getFastestRpcProvider(networkId: number) {
   const latencies: Record<string, number> = JSON.parse(localStorage.getItem("rpcLatencies") || "{}");
 
-  // Filter out latencies with a value of less than 0 because -1 means it failed
-  // Also filter out latencies that do not belong to the desired network
-  const validLatencies = Object.entries(latencies).filter(([key, latency]) => latency >= 0 && key.endsWith(`_${networkId}`));
+  for (let i = 0; i < MAX_RETRIES; ++i) {
+    // Filter out latencies with a value of less than 0 because -1 means it failed
+    // Also filter out latencies that do not belong to the desired network
+    const validLatencies = Object.entries(latencies).filter(([key, latency]) => latency >= 0 && key.endsWith(`_${networkId}`));
 
-  // Get all valid latencies from localStorage and find the fastest RPC
-  const sortedLatencies = validLatencies.sort((a, b) => a[1] - b[1]);
-  const optimalRPC = sortedLatencies[0][0].split("_").slice(0, -1).join("_"); // Remove the network ID from the key
+    // Get all valid latencies from localStorage and find the fastest RPC
+    const sortedLatencies = validLatencies.sort((a, b) => a[1] - b[1]);
+    const optimalRPC = sortedLatencies[0][0].split("_").slice(0, -1).join("_"); // Remove the network ID from the key
 
-  return new ethers.providers.JsonRpcProvider(optimalRPC, {
-    name: optimalRPC,
-    chainId: networkId,
-  });
+    try {
+      const rpcProvider = new ethers.providers.JsonRpcProvider(optimalRPC, {
+        name: optimalRPC,
+        chainId: networkId,
+      });
+      // We check if the networks positively gives us a block when requested to ensure the network works
+      // because some of them appear to be constantly failing such as https://gnosis.api.onfinality.io/public
+      await rpcProvider.getBlock(1);
+      return rpcProvider;
+    } catch (e) {
+      console.log(`Failed to get a block using network ${optimalRPC}, will try with another.`);
+      delete latencies[optimalRPC];
+    }
+  }
+
+  return null;
 }

--- a/static/scripts/rewards/rpc-optimization/get-optimal-provider.ts
+++ b/static/scripts/rewards/rpc-optimization/get-optimal-provider.ts
@@ -22,5 +22,10 @@ export async function useFastestRpc(app: AppState): Promise<JsonRpcProvider> {
     isTestCompleted = true;
   }
 
-  return getFastestRpcProvider(networkId);
+  const fastestProvider = await getFastestRpcProvider(networkId);
+  if (!fastestProvider) {
+    throw new Error("Failed to select a valid RPC Provider");
+  } else {
+    return fastestProvider;
+  }
 }

